sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,449 @@
+"""
+    SFF task manager
+    Deterministic link bring-up task manager for SFF compliant modules, running
+    as a thread inside xcvrd
+"""
+
+try:
+    import copy
+    import sys
+    import threading
+    import traceback
+
+    from swsscommon import swsscommon
+
+    from .xcvrd_utilities import port_mapping
+    from .xcvrd_utilities.xcvr_table_helper import XcvrTableHelper
+except ImportError as e:
+    raise ImportError(str(e) + " - required module not found")
+
+
+# Thread wrapper class for SFF compliant transceiver management
+
+
+class SffManagerTask(threading.Thread):
+
+    # CONFIG_DB port table fields:
+
+    ADMIN_STATUS = 'admin_status'
+    # This is the subport index for this logical port, starting from 1, 0 means
+    # all lanes are taken.
+    SUBPORT = 'subport'
+    # This is a comma separated list of lane numbers
+    LANES_LIST = 'lanes'
+
+
+    # STATE_DB TRANSCEIVER_INFO table fields:
+
+    # This filed can used to determine insertion/removal event. Since
+    # TRANSCEIVER_INFO has the same life cycle as a transceiver, if transceiver
+    # is inserted/removed, TRANSCEIVER_INFO is also created/deleted.
+    XCVR_TYPE = 'type'
+
+
+    # STATE_DB PORT_TABLE fields:
+
+    HOST_TX_READY = 'host_tx_ready'
+
+    # Default number of lanes per physical port for QSFP28/QSFP+ transceiver
+    DEFAULT_NUM_LANES_PER_PPORT = 4
+
+    # Subscribe to below tables in Redis DB
+    PORT_TBL_MAP = [
+        {
+            'CONFIG_DB': swsscommon.CFG_PORT_TABLE_NAME
+        },
+        {
+            'STATE_DB': 'TRANSCEIVER_INFO',
+            'FILTER': [XCVR_TYPE]
+        },
+        {
+            'STATE_DB': 'PORT_TABLE',
+            'FILTER': [HOST_TX_READY]
+        },
+    ]
+
+    def __init__(self, namespaces, main_thread_stop_event, platform_chassis, helper_logger):
+        threading.Thread.__init__(self)
+        self.name = "SffManagerTask"
+        self.exc = None
+        self.task_stopping_event = threading.Event()
+        self.main_thread_stop_event = main_thread_stop_event
+        self.helper_logger = helper_logger
+        self.platform_chassis = platform_chassis
+        # port_dict holds data obtained from on_port_update_event per port entry
+        # with logical_port_name as key.
+        # Its port entry will get deleted upon CONFIG_DB PORT_TABLE DEL.
+        self.port_dict = {}
+        # port_dict snapshot captured in the previous event update loop
+        self.port_dict_prev = {}
+        self.xcvr_table_helper = XcvrTableHelper(namespaces)
+        self.namespaces = namespaces
+
+    def log_notice(self, message):
+        self.helper_logger.log_notice("SFF: {}".format(message))
+
+    def log_warning(self, message):
+        self.helper_logger.log_warning("SFF: {}".format(message))
+
+    def log_error(self, message):
+        self.helper_logger.log_error("SFF: {}".format(message))
+
+    def get_active_lanes_for_lport(self, lport, subport_idx, num_lanes_per_lport, num_lanes_per_pport):
```

Comment:
@longhuan-cisco what are active lanes? is it the lanes mapped to the logical port? if so, can't we dervice from config DB's port table?